src/webextension/background.js

```diff
@@ -14,3 +14,92 @@ function openLockbox() {
 }
 
 browser.browserAction.onClicked.addListener(openLockbox);
+
+function makeEntrySummary(entry) {
+  return {site: entry.site, id: entry.id};
+}
+
+const fakeStore = {
+  _nextId: 0,
+  _datastore: [],
+
+  add(details) {
+    const entry = {...details, id: this._nextId++};
+    this._datastore.push(entry);
+    return entry;
+  },
+
+  update(entry) {
+    for (let i = 0; i < this._datastore.length; i++) {
+      if (this._datastore[i].id === entry.id) {
+        this._datastore[i] = entry;
+        break;
+      }
+    }
+    return entry;
+  },
+
+  remove(id) {
+    for (let i = 0; i < this._datastore.length; i++) {
+      if (this._datastore[i].id === id) {
+        this._datastore.splice(i, 1);
+        break;
+      }
+    }
+  },
+
+  get(id) {
+    for (let i = 0; i < this._datastore.length; i++) {
+      if (this._datastore[i].id === id) {
+        return this._datastore[i];
+      }
+    }
+    return null;
+  },
+
+  list() {
+    return this._datastore.map(makeEntrySummary);
+  },
+};
+
+let ports = new Set();
+browser.runtime.onConnect.addListener((port) => {
+  ports.add(port);
+  port.onDisconnect.addListener(() => {
+    console.log("message port disconnected");
+    ports.delete(port);
+  });
+});
+
+function broadcast(message) {
+  for (let p of ports)
+    p.postMessage(message);
+}
+
+browser.runtime.onMessage.addListener((message, sender, respond) => {
```

Comment:
since the real DataStore is async for just about everything, it would be interesting to see how that impacts this listener...

Reply:
Turns out all you have to do is return a Promise and it'll send that back as the response.